storage/rocksdb/clone/common.h

```diff
@@ -0,0 +1,703 @@
+    assert(is_snapshot());
+    assert(other.is_snapshot());
+
+    return id == other.id;
+  }
+
+  [[nodiscard]] bool has_applier_status() const noexcept {
+    return !applier_status.empty();
+  }
+
+  [[nodiscard]] bool is_invalid() const noexcept { return id == invalid_id; }
+
+  [[nodiscard]] bool is_version_check() const noexcept {
+    return id == version_check_id;
+  }
+
+  // Note that this only compares ID and ignores applier status, if any.
+  // Provided for std::unordered_map, otherwise better to use other methods most
+  // of the time.
+  [[nodiscard]] bool operator==(const locator &other) const noexcept {
+    return id == other.id;
+  }
+
+  void serialize(myrocks::Rdb_string_writer &buf) const {
+    assert(!is_invalid());
+    assert(buf.is_empty());
+
+    const auto payload_len =
+        total_fixed_length +
+        applier_status.size() * file_state::serialized_length;
+
+    assert(payload_len < std::numeric_limits<std::uint32_t>::max());
+    const metadata_header header{static_cast<std::uint32_t>(payload_len),
+                                 metadata_type::LOCATOR_V1};
+    header.serialize(buf);
+    buf.write_uint64(id);
+    for (const auto &file : applier_status) file.serialize(buf);
+
+    assert(buf.get_current_pos() == payload_len);
+  }
+
+  void serialize(myrocks::Rdb_string_writer &buf, const uchar *&loc,
+                 uint &loc_len) const {
+    serialize(buf);
+    loc = buf.ptr();
+    loc_len = buf.get_current_pos();
+  }
+
+  [[nodiscard]] static locator deserialize(const uchar *loc, uint loc_len) {
+    if (loc == nullptr || loc_len < total_fixed_length) return invalid();
+
+    auto buf = Rdb_string_reader{loc, loc_len};
+
+    const auto header = metadata_header::deserialize(buf);
+    if (header.is_invalid()) return invalid();
+    if (header.get_type() != metadata_type::LOCATOR_V1) return invalid();
+
+    const auto payload_len = header.get_payload_len();
+    if (payload_len < locator_fixed_length) return invalid();
+
+    const auto state_len = payload_len - total_fixed_length;
+    if ((state_len % file_state::serialized_length) != 0) return invalid();
+    const auto file_count = state_len / file_state::serialized_length;
+
+    std::uint64_t read_id;
+    if (buf.read_uint64(&read_id)) return invalid();
+
+    std::vector<file_state> read_applier_status;
+    read_applier_status.reserve(file_count);
+    for (unsigned i = 0; i < file_count; ++i) {
+      auto file_status = file_state::deserialize(buf);
+      if (file_status.is_invalid()) return invalid();
+      read_applier_status.push_back(std::move(file_status));
+    }
+
+    locator result{read_id, std::move(read_applier_status)};
+    assert(!result.is_invalid());
+    return result;
+  }
+
+ private:
+  static constexpr std::size_t locator_fixed_length = 8;
+  static constexpr std::size_t total_fixed_length =
+      metadata_header::length + locator_fixed_length;
+
+  static constexpr std::uint64_t version_check_id = 0;
```

Comment:
Could you add more comments about the locator structure and what the specific ids mean, i.e. 0 is reserved for the client check, valid locator IDs for snapshots are monotonically increasing from 1.